### PR TITLE
propagates more shape constraint attributes

### PR DESCRIPTION
**Pull request by wyzero**

### CODE REVIEW

**Comment by wyzero** [comment]
to #325

File: tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir [hunk 1]
```diff
@@ -165,6 +165,47 @@ func @main(%arg0: tensor<?x?x?xf32>, %arg1: tensor<?x?xf32>) -> tensor<?x?xf32>
 
 // -----
 
+// Test arith.cmpi + arith::muli + arith::add
+// CHECK-LABEL: main
+func @main(%arg0: tensor<?x?x?xf32>) -> (index, index) {
```
Comment (fortianyou) [inline tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir:170]: Nit: split the test into simple and tiny cases, one unit test one thing

File: tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir [hunk 2]
```diff
@@ -165,6 +165,47 @@ func @main(%arg0: tensor<?x?x?xf32>, %arg1: tensor<?x?xf32>) -> tensor<?x?xf32>
...
+// CHECK-LABEL: main
+func @main(%arg0: tensor<?x?x?xf32>, %arg1: tensor<2xindex>) -> (tensor<?x?xf32>, index) {
+  // CHECK: %[[C0:.*]] = arith.constant 0 : index
+  // CHECK: return %[[RET0:.*]], %[[C0]]
+  %c-1 = arith.constant -1 : index
+  %c1 = arith.constant 1 : index
+  %c0 = arith.constant 0 : index
+  %0 = "mhlo.dynamic_reshape"(%arg0, %arg1) : (tensor<?x?x?xf32>, tensor<2xindex>) -> tensor<?x?xf32>
+  %1 = tensor.extract %arg1[%c0] : tensor<2xindex>
+  %2 = arith.cmpi eq, %1, %c-1 : index
```
Comment (fortianyou) [inline tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir:202]: Is that mean the `%arg1` elements can't be `-1`?

Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir:202]: `%arg1` is used as the target shape in  `%0 = "mhlo.dynamic_reshape"(%arg0, %arg1)`, thus we can infer that all the elements of `%arg1` should not be `-1`.

File: tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir (hunk 1 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir:170]: Actually, I did this by design. It was designed to check the chain of propagation is ok.

**Review by fortianyou — APPROVED** [review]
LGTM

File: tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir (hunk 2 above)
Comment (fortianyou) [reply tao_compiler/mlir/disc/transforms/tests/disc-shape-optimization.mlir:202]: I see.

**Pull request merged**